extern/boostd-data/couchbase/db.go

```diff
@@ -665,3 +665,93 @@ func serviceName(svc gocb.ServiceType) string {
 	}
 	return "unknown"
 }
+
+// RemoveMetadata
+func (db *DB) RemoveMetadata(ctx context.Context, pieceCid cid.Cid) error {
+	ctx, span := tracing.Tracer.Start(ctx, "db.remove_metadata")
+	defer span.End()
+
+	k := toCouchKey(pieceCid.String())
+	metadata, err := db.GetPieceCidToMetadata(ctx, pieceCid)
+	if err != nil {
+		if errors.Is(err, gocb.ErrDocumentNotFound) {
+			return nil
+		}
+		return fmt.Errorf("getting piece %s metadata: %w", pieceCid, err)
+	}
+
+	// Remove all multihashes first, as without Metadata, they cannot be removed.
+	// This order is important as metadata.BlockCount is required in case RemoveIndexes fails
+	// and needs to be run manually
+	if err = db.RemoveIndexes(ctx, pieceCid, metadata.BlockCount); err != nil {
+		return fmt.Errorf("failed removing index for piece %s: %w", pieceCid, err)
+	}
+
+	_, err = db.pcidToMeta.Remove(k, &gocb.RemoveOptions{Context: ctx})
+	if err != nil {
+		return fmt.Errorf("removing piece %s metadata: %w", pieceCid, err)
+	}
+
+	return nil
+}
+
+// RemoveIndexes
+func (db *DB) RemoveIndexes(ctx context.Context, pieceCid cid.Cid, recordCount int) error {
+	ctx, span := tracing.Tracer.Start(ctx, "db.remove_indexes")
+	defer span.End()
+
+	// Get the number of shards
+	_, totalShards := getShardPrefixBitCount(recordCount)
+	for i := 0; i < totalShards; i++ {
+		// Get the map of multihash -> offset/size for the shard
+		shardPrefix, err := getShardPrefix(i)
+		if err != nil {
+			return err
+		}
+		cbKey := toCouchKey(pieceCid.String() + shardPrefix)
+		cbMap := db.pieceOffsets.Map(cbKey)
+		recMap, err := cbMap.Iterator()
+		if err != nil {
+			if isNotFoundErr(err) {
+				// If there are no records in a particular shard just skip the shard
+				continue
+			}
+			return fmt.Errorf("getting all records for piece %s: %w", pieceCid, err)
+		}
+
+		// Get each value in the map
+		for mhStr := range recMap {
+			cbKey := toCouchKey(mhStr)
+			getResult, err := db.mhToPieces.Get(cbKey, &gocb.GetOptions{Context: ctx})
+			if err != nil {
+				return nil
+			}
+			var cidStrs []string
+			err = getResult.Content(&cidStrs)
+			if err != nil {
+				return err
+			}
+
+			// Remove multihash -> pieceCId (value only)
+			for i, v := range cidStrs {
+				if v == pieceCid.String() {
+					cidStrs[i] = cidStrs[len(cidStrs)-1]
+					cidStrs[len(cidStrs)-1] = ""
+					cidStrs = cidStrs[:len(cidStrs)-1]
+				}
+			}
+
+			_, err = db.mhToPieces.Replace(cbKey, cidStrs, &gocb.ReplaceOptions{Context: ctx})
+			if err != nil {
+				return fmt.Errorf("removing multihash %s to piece %s: update doc: %w", mhStr, pieceCid, err)
+			}
```

Comment:
This block needs to be wrapped in a compare-and-swap. See AddDealForPiece for an example.
You can read more about compare-and-swap at https://docs.couchbase.com/go-sdk/current/howtos/concurrent-document-mutations.html